### PR TITLE
Rewrite README.md and CONTRIBUTING to match the ones from 'cucumber-ruby'

### PR DESCRIPTION
**Pull request by aurelien-reeves**
**Is your pull request related to a problem? Please describe.**

https://github.com/cucumber/common/issues/1521

**Describe the solution you have implemented**

I have adopted the formalism from cucumber-ruby issued from https://github.com/cucumber/cucumber-ruby/pull/1542

### CODE REVIEW

File: docs/ARCHITECTURE.md [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+### The AST
+
+The Abstract Syntax Tree or AST is an object graph that represents the Gherkin documents you've passed into the core. Things like `Feature`, `Scenario` and `ExamplesTable`.
+
+These are immutable value objects.
+
+### Test cases
+
+Your gherkin might contain scenarios, as well as examples from tables beneath a scenario outline.
```
Comment (olleolleolle) [inline docs/ARCHITECTURE.md:21]: This could be capitalized, I guess, like on the line before.

File: docs/ARCHITECTURE.md [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+Test cases represent the general case of both of these. We compile the AST down to instances of `Cucumber::Core::Test::Case`, each containing a number of instances of `Cucumber::Core::Test::Step`. It's these that are then filtered and executed.
+
+Test cases and their test steps are also immutable value objects.
```
Comment (olleolleolle) [inline docs/ARCHITECTURE.md:25]: Does this sentence work without the "also"? If yes, would that read better?

File: docs/ARCHITECTURE.md (hunk 1 above)
Comment (aurelien-reeves) [reply docs/ARCHITECTURE.md:21]: You mean "Gherkin"? Yes, I'll fix this :)

File: docs/ARCHITECTURE.md (hunk 2 above)
Comment (aurelien-reeves) [reply docs/ARCHITECTURE.md:25]: I guess so

I copy/paste that from the actual README. If an native english speaker could confirm
@mattwynne

**Comment by aurelien-reeves** [comment]
It looks a lot like the cucumber-ruby README, CONTRIBUTING and PULL_REQUEST_TEMPLATE documents, but there are less info when it comes to:

- contributing to something else than the code of cucumber-ruby-core: there is a "please see cucumber-ruby" for that
- needs for updating docs.cucumber.io in PR Template: it seems to be not relevant for cucumber-core, such needs for docs.cucumber.io updates would be related to changes in cucumber-ruby

File: README.md [hunk 3]
```diff
@@ -1,123 +1,131 @@
...
 require 'cucumber/core/filter'
 
-# This is the most complex part of the example. The filter takes test cases as input,
-# activates each step with an action block, then passes a new test case with those activated
-# steps in it on to the next filter in the chain.
 class ActivateSteps < Cucumber::Core::Filter.new
   def test_case(test_case)
     test_steps = test_case.test_steps.map do |step|
-      activate(step)
+      step.with_action { puts "processing '#{step.text}'" }
```
Comment (luke-hill) [inline README.md:78]: We should probably be more specific and say `Kernel.puts` here just so we don't conflict with any deprecations from cucumber 5/6

File: docs/ARCHITECTURE.md [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+### Filters
+
+Once we have the test cases, and they've been activated by the mappings, you may want to pass them through a filter or two. Filters can be used to do things like activate, sort, replace or remove some of the test cases or their steps before they're executed.
+
+### Events
+
+Events are how you find out what is happening during your test run. As the test cases and steps are executed, the runner emits events to signal what's going on.
+
+The following events are emitted during a run:
```
Comment (luke-hill) [inline docs/ARCHITECTURE.md:35]: The following, is a non-exhaustive list of the events that are emitted during a run

OR

Some of the events that can be emitted during a run are:

**Review by luke-hill — COMMENTED** [review]
Seems solid, couple of minor language tweaks.

**Comment by aurelien-reeves** [comment]
> 
> 
> Seems solid, couple of minor language tweaks.

Thanks,

Fixed

**Review by luke-hill — APPROVED** [review]

**Pull request merged**